Mathlib/Algebra/Order/Hom/Basic.lean

```diff
@@ -118,3 +137,157 @@ theorem le_map_div_add_map_div [Group α] [AddCommSemigroup β] [LE β] [MulLEAd
+attribute [simp] map_neg_eq_map
+attribute [simp] map_inv_eq_map
+
+-- See note [lower instance priority]
+instance (priority := 100) AddGroupSeminormClass.toZeroHomClass [AddGroup α]
+  [OrderedAddCommMonoid β] [AddGroupSeminormClass F α β] : ZeroHomClass F α β :=
+{ ‹AddGroupSeminormClass F α β› with }
+
+section GroupSeminormClass
+variable [Group α] [OrderedAddCommMonoid β] [GroupSeminormClass F α β] (f : F) (x y : α)
+
+@[to_additive] lemma map_div_le_add : f (x / y) ≤ f x + f y :=
+by rw [div_eq_mul_inv, ←map_inv_eq_map f y]; exact map_mul_le_add _ _ _
+
+@[to_additive] lemma map_div_rev : f (x / y) = f (y / x) := by rw [←inv_div, map_inv_eq_map]
+
+@[to_additive] lemma le_map_add_map_div' : f x ≤ f y + f (y / x) :=
+by simpa only [add_comm, map_div_rev, div_mul_cancel'] using map_mul_le_add f (x / y) y
+
+end GroupSeminormClass
+
+example [OrderedAddCommGroup β] : OrderedAddCommMonoid β := inferInstance
+
+@[to_additive] lemma abs_sub_map_le_div [Group α] [LinearOrderedAddCommGroup β]
+  [GroupSeminormClass F α β] (f : F) (x y : α) : |f x - f y| ≤ f (x / y) := by
+  rw [abs_sub_le_iff, sub_le_iff_le_add', sub_le_iff_le_add']
+  exact ⟨le_map_add_map_div _ _ _, le_map_add_map_div' _ _ _⟩
+
+@[to_additive] -- See note [lower instance priority]
+instance (priority := 100) GroupSeminormClass.toNonnegHomClass [Group α]
+  [LinearOrderedAddCommMonoid β] [GroupSeminormClass F α β] :
+  NonnegHomClass F α β :=
+{ ‹GroupSeminormClass F α β› with
+  map_nonneg := fun f a ↦ (nsmul_nonneg_iff two_ne_zero).1 $
+    by rw [two_nsmul, ←map_one_eq_zero f, ←div_self' a]; exact map_div_le_add _ _ _ }
+
+section GroupNormClass
+variable [Group α] [OrderedAddCommMonoid β] [GroupNormClass F α β] (f : F) {x : α}
+
+@[simp, to_additive] lemma map_eq_zero_iff_eq_one : f x = 0 ↔ x = 1 :=
+⟨eq_one_of_map_eq_zero _, by rintro rfl; exact map_one_eq_zero _⟩
+
+@[to_additive] lemma map_ne_zero_iff_ne_one : f x ≠ 0 ↔ x ≠ 1 := (map_eq_zero_iff_eq_one _).not
+
+end GroupNormClass
+
+-- This instance is simply the composition of the instances
+-- `GroupSeminormClass.toNonnegHomClass` and `GroupNormClass.toGroupSeminormClass`.
+-- However, it has to be declared explicitly due to hitting an edge case in the inference algorithm:
+-- those instances both depend on a different subclass of `OrderedAddMonoid β`, but `β` is an
+-- `out_param` that cannot be inferred until both instances are applied.
+-- So we have to do it manually until Lean becomes smarter.
+-- See note [lower instance priority]
+@[to_additive] instance (priority := 100) GroupNormClass.toNonnegHomClass [Group α]
+  [LinearOrderedAddCommMonoid β] [GroupNormClass F α β] : NonnegHomClass F α β :=
+GroupSeminormClass.toNonnegHomClass
+
+@[to_additive] lemma map_pos_of_ne_one [Group α] [LinearOrderedAddCommMonoid β]
+  [GroupNormClass F α β] (f : F) {x : α} (hx : x ≠ 1) : 0 < f x :=
+(@map_nonneg F α β _ _ GroupSeminormClass.toNonnegHomClass f x).lt_of_ne
+  ((map_ne_zero_iff_ne_one _).2 hx).symm
```

Comment:
@Vierkantor, this is even worse that the previous. Even when I give the exact instance just above, Lean fails to find it.

Reply:
This is because Lean 4 handles `outParam`s somewhat differently: if you check the dangerousInstances linter, you'll see that there are a lot of errors. You'll need to modify these instances such as `GroupNormClass.toNonnegHomClass` so that the instance parameters `[Group α] [LinearOrderedAddCommMonoid β]` that depend only on `outParam`s must become implicit parameters `{_ : Group α} {_ : LinearOrderedAddCommMonoid β}` instead.

Reply:
Oh this is going to be a headache.